tools/SampleAnalyzer/Process/Counter/Counter.h

```diff
@@ -1,114 +1,128 @@
 //  GNU General Public License for more details.
-//  
+//
 //  You should have received a copy of the GNU General Public License
 //  along with MadAnalysis 5. If not, see <http://www.gnu.org/licenses/>
-//  
+//
 ////////////////////////////////////////////////////////////////////////////////
 
-
 #ifndef COUNTER_h
 #define COUNTER_h
 
-
 // SampleAnalyzer headers
 #include "SampleAnalyzer/Commons/Base/PortableDatatypes.h"
+#include "SampleAnalyzer/Commons/Base/Basics.h"
+#include "SampleAnalyzer/Commons/DataFormat/WeightCollection.h"
 
 // STL headers
 #include <iostream>
 #include <string>
 #include <map>
 
-
 namespace MA5
 {
+    class CounterManager;
 
-class CounterCollection;
-
-class Counter
-{
-  friend class CounterCollection;
-
-  // -------------------------------------------------------------
-  //                        data members
-  // -------------------------------------------------------------
- public :
-
-  /// name of the analysis
-  std::string name_;
-
-  /// number of times the function Increment is called
-  /// first = positive weight ; second = negative weight
-  std::pair<MAint64,MAint64> nentries_;
-
-  /// sum of weights
-  /// first = positive weight ; second = negative weight
-  std::pair<MAfloat64,MAfloat64> sumweight_;
-
-  /// sum of squared weights
-  /// first = positive weight ; second = negative weight
-  std::pair<MAfloat64,MAfloat64> sumweight2_;
-
-
-  // -------------------------------------------------------------
-  //                       method members
-  // -------------------------------------------------------------
- public :
-
-  /// Constructor without argument 
-  Counter(const std::string& name = "unkwown")
-  { 
-    name_       = name;
-    nentries_   = std::make_pair(0,0); 
-    sumweight_  = std::make_pair(0.,0.);
-    sumweight2_ = std::make_pair(0.,0.);
-  }
-
-  /// Destructor
-  ~Counter()
-  { }
-
-  /// Reset
-  void Reset()
-  {
-    nentries_   = std::make_pair(0,0); 
-    sumweight_  = std::make_pair(0.,0.);
-    sumweight2_ = std::make_pair(0.,0.);
-  }
-
-  /// Increment the counter
-  void Increment(const MAfloat32& weight=1.)
-  {
-    if (weight>0)
-    {
-      nentries_.first++;
-      sumweight_.first+=weight;
-      sumweight2_.first+=weight*weight;
-    }
-    else if (weight<0)
+    class Counter
     {
-      nentries_.second++;
-      sumweight_.second+=weight;
-      sumweight2_.second+=weight*weight;
-    }
-  }
-
-};
+        friend class CounterManager;
+
+        // -------------------------------------------------------------
+        //                        data members
+        // -------------------------------------------------------------
+    public:
+        /// name of the analysis
+        std::string name_;
+
+        /// number of times the function Increment is called
+        /// first = positive weight ; second = negative weight
+        std::map<MAint32, ENTRIES> nentries_;
+
```

Comment:
I think that this is incorrect. `nentries_` should be a `std::pair<MAint32,MAint32>`. If there is an entry, then there is an entry for every single weight. It is not necessary to have an `nentries_` element per weight.

Reply:
But weights can be different i.e. positive or negative so it will depend on weight

Reply:
Yes yes... I overlooked one option :)